framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,597 @@
+import javax.lang.model.type.TypeMirror;
+
+import org.checkerframework.dataflow.cfg.node.FieldAccessNode;
+import org.checkerframework.dataflow.cfg.node.ImplicitThisLiteralNode;
+import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.dataflow.cfg.node.ReturnNode;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedArrayType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedTypeVariable;
+import org.checkerframework.framework.util.AnnotatedTypes;
+import org.checkerframework.framework.util.AnnotationBuilder;
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.InternalUtils;
+
+import annotations.Annotation;
+import annotations.el.AClass;
+import annotations.el.AField;
+import annotations.el.AMethod;
+import annotations.el.AScene;
+import annotations.el.ATypeElement;
+import annotations.el.AnnotationDef;
+import annotations.el.DefException;
+import annotations.el.InnerTypeLocation;
+import annotations.field.AnnotationFieldType;
+import annotations.field.ArrayAFT;
+import annotations.field.BasicAFT;
+import annotations.field.ScalarAFT;
+import annotations.io.IndexFileParser;
+import annotations.io.IndexFileWriter;
+import annotations.util.JVMNames;
+
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Attribute.Array;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Symbol.TypeSymbol;
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.TypeAnnotationPosition;
+import com.sun.tools.javac.code.Type.ArrayType;
+import com.sun.tools.javac.code.Type.ClassType;
+
+/**
+ * SignatureInferenceScenes represents a set of annotations that are inferred
+ * in a program. It also writes those annotations into .jaif files.
+ * This class infers annotations for fields, method return types, and method
+ * parameters.
+ * The set of annotations inferred for a certain class is stored in an
+ * {@link annotations.el.AScene}, which will later be written into a .jaif file.
+ * For example, a class field of a class whose fully-qualified name is
+ * my.package.MyClass will have its inferred type stored in a Scene, and later
+ * written into a file named my.package.MyClass.jaif.
+ *
+ * Calling an update* method
+ * ({@link updateInferredFieldType}, {@link updateInferredMethodParametersTypes}, or
+ * {@link updateInferredMethodReturnType}) 
+ * reads the currently-stored type for an element in a Scene, if any,
+ * and replaces it by the LUB of it and the update method's argument.
+ *
```

Comment:
This summary of the behavior of the class does not say anything about reading existing .jaif files, but the class does so.  The documentation for methods like `updateInferredMethodParametersTypes` doesn't say anything about reading existing files, either.  Can you more fully explain what the class does?